requirements/base.in

```diff
@@ -8,6 +8,7 @@ parameterized
 pre-commit
 pylint
 pytest
+pytest-mock
```

Comment:
I think this is an important change made by this PR. If we add this dependency, we should try to use it everywhere for consistency. It's the first time I see it, so could you please explain why you think this is better than using `unittest.mock` directly? Thanks!

Reply:
> With this I found a way to implement multiple mocks on a single fixture to the passed them to several test cases.

Do you have a specific case where this was resulting to be tricky without the library? So that we can check the actual scenario.

Reply:
It is not that is was tricky. It could have been achieved. However, without the `mocker` plugin, I found mocking and patching with `mock.patch` repetitive and unclean. Maybe because the way I would've implemented the solutions with it were so.

For example, I mocked at the beginning of test_aggregator.py all of the external connections in a fixture. As an advantage, mocker allows to patch/mock a function inside a fixture  so that when using the fixture in any other method, this patches would be automatically implemented. 
mock.patch-ing in a fixture however is not implemented and that would force to patch somewhere else (inside the test case or as a decorator).
In contrast, patching external connections for so many methods (with mock.patch), would end up with several stacked patch decorators before each one of them (or the classes grouping them), or with several nested patch context managers. 
I know the first one could be done just once if all of the test cases were grouped in one single class. However, due to the differences between methods being tested, other patches would be also needed for some of the methods, being some times several, and again, repeated over various test cases, causing repetition, and as I read, repeating a lot when coding is not a good practice, specially when it gets to testing due to the maintenance of these tests.

I did not found any further advantage on using mocker over mock.patch other than it helps out to get a cleaner implementation. If you think adding this dependency is not worth it, I can redesign the tests just using `mock.patch `

Reply:
I do want to mention something that I find weird.

The main reason I added this dependency to the file, is because when tox ran, it could not ran the tests precisely because this pytest-mock was apparently missing. However, in my machine I never had to install anything additional other than that installed when running `$ pip install -e .` right after cloning the repo.

This makes me think that the virtual environment created when tox runs has some difference with that one built when running these steps [here](https://gridsingularity.atlassian.net/wiki/spaces/D3A/pages/2243461121/How+to+set+up+local+development+environment+macOS#GSy-e-SDK).

I'm not sure if that would make sense though. I just comment this because adding this dependency might not be needed at all afterall.

Reply:
> patching external connections for so many methods (with mock.patch), would end up with several stacked patch decorators before each one of them (or the classes grouping them), or with several nested patch context managers.

In order to achieve something similar you are doing now without using mocker, you could put the patches into a list and start and stop the patches in the `setup_method` and `teardown_method`, like here:
https://github.com/gridsingularity/gsy-web/blob/master/simulation/tests/test_simulation_data_handler.py

I am not strictly against using mocker, but Pierpaolo has a point with the consistency. If you are still not convinced, please keep using it.

Reply:
@hannesdiedrich I believe that could actually work as an alternative to patch without using `mocker`. I could implement it so that we get rid of the plugin. I'm not sure I would implement a `setup_method` though, I would maybe use `yield` statement inside the fixtures to set up that tear down logic. That is mainly because I have already implemented each test case to use the exact patching-fixture it "needs" and writing a set-up and tear-down in each of those classes, would cause again that some test cases use unnecessary patches.

I ask from you as reviewers your opinion on what would be best before I get to it. If you think it is not necessary due to the conclusion reached on slack, please say as well.

Reply:
@ERafaelMartinez Please correct me if I'm wrong: the PR is quite big and I might miss something. I take one example: in the `fixture_mock_connections` fixture, you patch the `retrieve_jwt_key_from_server` method. This fixture is then applied to all tests. However, `retrieve_jwt_key_from_server` is not used in all tests. For example, the `start_websocket_connection()` method doesn't need to patch `retrieve_jwt_key_from_server()`, because that's never reached in that method.
Can you confirm this, or am I mistaken here?

Reply:
Oh wait, I think I'm wrong here! I've seen now that the `__init__()` is inheriting from `RestAssetClient` and that's where that method is called. Apologies for missing that part!